### DISCUSSION TITLE
How to use main.cpp in example folder

### DISCUSSION THREAD

**Discussion by Showwwwwwwww** [ggml-org/llama.cpp#3281] [Q&A]
Hi, 

Can I know how to use the main.cpp file which stored in example folder? I tried to modify the readline function in the main.cpp which allowed the llama chat with a txt file instead of console input. Sorry I am barely use c++ and have no idea for how to use it.

Thanks.

**Comment by goerch** [comment]
You have three [supported options](https://github.com/ggerganov/llama.cpp#build), otherwise you are on your own.

I'd personally recommend using `cmake`.

**Reply by Showwwwwwwww** [reply]
I tried to use cmake to compile the **main.cpp** which stored in the example file. 
**However it shows the error message**
`main % mkdir build
cd build
cmake ..
cmake --build . --config Release
CMake Warning (dev) in CMakeLists.txt:
  No project() command is present.  The top-level CMakeLists.txt file must
  contain a literal, direct call to the project() command.  Add a line of
  code such as
project(ProjectName)

  near the top of the file, but after cmake_minimum_required().

  CMake is pretending there is a "project(Project)" command on the first
  line.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in CMakeLists.txt:
  cmake_minimum_required() should be called prior to this top-level project()
  call.  Please see the cmake-commands(7) manual for usage documentation of
  both commands.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- The C compiler identification is AppleClang 14.0.3.14030022
-- The CXX compiler identification is AppleClang 14.0.3.14030022
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Warning (dev) in CMakeLists.txt:
  No cmake_minimum_required command is present.  A line of code such as

    cmake_minimum_required(VERSION 3.27)

  should be added at the top of the file.  The version specified may be lower
  if you wish to support older CMake versions for this project.  For more
  information run "cmake --help-policy CMP0000".
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Configuring done (0.5s)
-- Generating done (0.0s)
-- Build files have been written to: /Users/chen/Desktop/robot/robot_research/llama2/llama.cpp/examples/main/build
[ 50%] Building CXX object CMakeFiles/main.dir/main.o
/Users/chen/Desktop/robot/robot_research/llama2/llama.cpp/examples/main/main.cpp:6:10: fatal error: 'common.h' file not found
#include "common.h"
         ^~~~~~~~~~
1 error generated.
make[2]: *** [CMakeFiles/main.dir/main.o] Error 1
make[1]: *** [CMakeFiles/main.dir/all] Error 2
make: *** [all] Error 2
`

**Reply by goerch** [reply]
That looks like the correct command sequence:

```
mkdir build
cd build
cmake ..
cmake --build . --config Release
``` 

The documentation maybe should state that you have to run these in your `llama.cpp` project folder, i.e. the folder containing [this](https://github.com/ggerganov/llama.cpp/blob/master/CMakeLists.txt) file.
